File: Source/Aladdin.CAPI.ANSI/CAPI/ANSI/Cipher/DESX.cs
using System;

namespace Aladdin.CAPI.ANSI.Cipher
{
    ///////////////////////////////////////////////////////////////////////////
    // Алгоритм шифрования DES-X
    ///////////////////////////////////////////////////////////////////////////
    public class DESX : BlockCipher
    {
        // конструктор
        public DESX(CAPI.Factory factory, SecurityStore scope) : base(factory, scope) {}

        // тип ключа
        public override SecretKeyFactory KeyFactory { get { return Keys.DESX.Instance; }}
        // размер блока
        public override int BlockSize { get { return 8; }}

        // получить режим шифрования
        public override CAPI.Cipher CreateBlockMode(CipherMode mode)
        {
            // в зависимости от режима
            if (mode is CipherMode.ECB) 
            {
                // закодировать параметры алгоритма
                ASN1.IEncodable parameters = ASN1.Null.Instance; 

                // получить алгоритм шифрования
                using (CAPI.Cipher engine = Factory.CreateAlgorithm<CAPI.Cipher>(
                    Scope, ASN1.ANSI.OID.ssig_des_ecb, parameters)) 
                {
                    // вернуть алгоритм шифрования
                    if (engine == null) throw new NotSupportedException();
                
                    // создать модификацию алгоритма
                    using (CAPI.Cipher desX = new Engine.DESX(engine))
                    {
                        // вернуть режим алгоритма
                        return new BlockMode.PaddingConverter(desX, PaddingMode.Any); 
                    }
                }
            }
            // в зависимости от режима
            if (mode is CipherMode.CBC) 
            {
                // закодировать параметры алгоритма
                ASN1.IEncodable parameters = new ASN1.OctetString(((CipherMode.CBC)mode).IV); 

                // получить алгоритм шифрования
                using (CAPI.Cipher cipher = Factory.CreateAlgorithm<CAPI.Cipher>(
                    Scope, ASN1.ANSI.OID.rsa_desx_cbc, parameters))
                {
                    // проверить наличие алгоритма
                    if (cipher != null) return cipher; 
                }
            }
            // вызвать базовую функцию
            return CreateBlockMode(mode, 24); 
        }
    }
}

File: Source/Aladdin.CAPI.ANSI.Java/src/aladdin/capi/ansi/cipher/DESX.java
package aladdin.capi.ansi.cipher;
import aladdin.asn1.*;
import aladdin.asn1.ansi.*;
import aladdin.capi.*; 
import java.io.*; 

///////////////////////////////////////////////////////////////////////////
// Алгоритм шифрования DES-X
///////////////////////////////////////////////////////////////////////////
public final class DESX extends BlockCipher
{
    // конструктор
    public DESX(Factory factory, SecurityStore scope) { super(factory, scope); }
    
    // тип ключа
    @Override public final SecretKeyFactory keyFactory() 
    { 
        // тип ключа
        return aladdin.capi.ansi.keys.DESX.INSTANCE; 
    } 
    // размер блока
    @Override public final int blockSize() { return 8; } 
        
    // получить режим шифрования
    @Override public Cipher createBlockMode(CipherMode mode) throws IOException
    {
        // в зависимости от режима
        if (mode instanceof CipherMode.ECB) 
        {
            // закодировать параметры алгоритма
            IEncodable parameters = Null.INSTANCE; 
            
            // получить алгоритм шифрования
            try (Cipher engine = (Cipher)factory().createAlgorithm(
                scope(), OID.SSIG_DES_ECB, parameters, Cipher.class)) 
            {
                // вернуть алгоритм шифрования
                if (engine == null) throw new UnsupportedOperationException();
                
                // создать модификацию алгоритма
                try (Cipher desX = new aladdin.capi.ansi.engine.DESX(engine))
                {
                    // вернуть режим алгоритма
                    return new BlockMode.PaddingConverter(desX, PaddingMode.ANY); 
                }
            }
        }
        // в зависимости от режима
        if (mode instanceof CipherMode.CBC) 
        {
            // закодировать параметры алгоритма
            IEncodable parameters = new OctetString(((CipherMode.CBC)mode).iv()); 
            
            // получить алгоритм шифрования
            Cipher cipher = (Cipher)factory().createAlgorithm(
                scope(), OID.RSA_DESX_CBC, parameters, Cipher.class
            ); 
            // проверить наличие алгоритма
            if (cipher != null) return cipher; 
        }
        // вызвать базовую функцию
        return createBlockMode(mode, 24); 
	}
}
